diff --git a/nikki/Makefile b/nikki/Makefile
--- a/nikki/Makefile
+++ b/nikki/Makefile
@@ -72,6 +72,7 @@ define Package/nikki/install
 
 	$(INSTALL_BIN) $(CURDIR)/files/scripts/include.sh $(1)/etc/nikki/scripts/include.sh
 	$(INSTALL_BIN) $(CURDIR)/files/scripts/firewall_include.sh $(1)/etc/nikki/scripts/firewall_include.sh
+	$(INSTALL_BIN) $(CURDIR)/files/scripts/dns_gateway.sh $(1)/etc/nikki/scripts/dns_gateway.sh
 	$(INSTALL_BIN) $(CURDIR)/files/scripts/debug.sh $(1)/etc/nikki/scripts/debug.sh
 
 	$(INSTALL_BIN) $(CURDIR)/files/nftables/geoip_cn.nft $(1)/etc/nikki/nftables/geoip_cn.nft
diff --git a/nikki/files/nikki.init b/nikki/files/nikki.init
index 5bc88777eb02e88317d7f5afd160c01261b6aab4..ff5c8bd12c74b5784e2b5b896645308f0099bc2a 100644
--- a/nikki/files/nikki.init
+++ b/nikki/files/nikki.init
@@ -405,6 +405,13 @@ service_started() {
 			log "Proxy" "Hijack failed."
 			log "App" "Exit."
 		fi
+		# dns gateway
+		$DNS_GATEWAY_SH apply
+		if nft list tables | grep -q dns_gateway; then
+			log "DNS" "Gateway ready."
+		else
+			log "DNS" "Gateway setup failed."
+		fi
 	fi
 }
 
@@ -455,6 +462,7 @@ cleanup() {
 	for handle in $handles; do
 		nft delete rule inet fw4 forward handle "$handle"
 	done
+	$DNS_GATEWAY_SH cleanup
 	# delete started flag
 	rm "$STARTED_FLAG_PATH" > /dev/null 2>&1
 	# revert fix compatible with dockerd
diff --git a/nikki/files/scripts/include.sh b/nikki/files/scripts/include.sh
index bff99a376bc08a156ffd79be887644151bd898f5..0e29f685117169d9950f6500b1896eeaa1139aff 100644
--- a/nikki/files/scripts/include.sh
+++ b/nikki/files/scripts/include.sh
@@ -33,6 +33,7 @@ HIJACK_UT="$UCODE_DIR/hijack.ut"
 SH_DIR="$HOME_DIR/scripts"
 INCLUDE_SH="$SH_DIR/include.sh"
 FIREWALL_INCLUDE_SH="$SH_DIR/firewall_include.sh"
+DNS_GATEWAY_SH="$SH_DIR/dns_gateway.sh"
 
 # nftables
 NFT_DIR="$HOME_DIR/nftables"
